### PR TITLE
[JMENanoAOD] Updates for Run-3

### PR DESCRIPTION
**Pull request by nurfikri89**
#### PR description:
This PR updates JMENano to be compatible with Run-3 data and MC, starting with the upcoming NanoAODv10 production. The following major changes are made: 

- `Sequences` have been changed to `Tasks` and JMENano can now run in Prompt.
- Updated names of NanoAOD-related config files and modules.
- Add a dedicated function for reclustering and content customisation of AK4 Puppi jets.
- Recluster both AK4 CHS and AK4 Puppi jet collections but modify their table names and content depending on era (i.e Run-2 or Run-3).
- Increase pt threshold for stored reclustered AK8 reco jets to 15 GeV.
- The JMENano workflows are enabled back again, which was disabled by PR #35140. Enable Run-3 workflows and added one wf: 11634.15.

#### PR validation:
- passes the (re-enabled) JMENano workflows: `runTheMatrix.py -i all --ibeos -l 10224.15,11024.15,25202.15,11634.15`

With this PR, the event size is 6.24 kb/event. Test was done using 10K events from a TTTo2J1L1Nu Run3Winter22MiniAOD file. Below is the CPU time:

jmenano_cputtime

The full content and size details can be found in the JMENano twiki: [content](https://twiki.cern.ch/twiki/pub/CMS/JMECustomNanoAOD/JMENano_version4_forPR_39000_content.html) / [size](https://twiki.cern.ch/twiki/pub/CMS/JMECustomNanoAOD/JMENano_version4_forPR_39000_size.html)

#### if this PR is a backport please specify the original PR and why you need to backport that PR:
Once merged, this PR needs to be backported to 12_4_X for the JMENanoAODv10 production.

### CODE REVIEW

**Comment by mariadalfonso** [comment]
please test

**Comment by mariadalfonso** [comment]
1. size 6.33 kb/event is within the allocated budget
2. RecoBTag takes a lot of time, in the future should be optimized @hqucms maybe can have a look later
3. 10224.15,11024.15,25202.15 which one is Run3 ?
4. What is the difference between the  
FatJet and  FatJetForJEC 
https://twiki.cern.ch/twiki/pub/CMS/JMECustomNanoAOD/JMENano_version4_forPR_39000_content.html#FatJetForJEC)
https://twiki.cern.ch/twiki/pub/CMS/JMECustomNanoAOD/JMENano_version4_forPR_39000_content.html#FatJet

File: PhysicsTools/NanoAOD/python/custom_jme_cff.py [hunk 1]
```diff
@@ -762,6 +899,59 @@ def ReclusterAK4CHSJets(proc, recoJA, runOnMC):
...
+  (~run2_nanoAOD_ANY).toModify(proc.updatedJetsWithUserData.userFloats, leptonPtRatiov0 = None)
+  (~run2_nanoAOD_ANY).toModify(proc.updatedJetsWithUserData.userFloats, leptonPtRelInvv0 = None)
+  (~run2_nanoAOD_ANY).toModify(proc.updatedJetsWithUserData.userFloats, leptonDeltaR = None)
+  (~run2_nanoAOD_ANY).toModify(proc.updatedJetsWithUserData.userFloats, leptonPt = None)
+  (~run2_nanoAOD_ANY).toModify(proc.updatedJetsWithUserData.userFloats, vtxPt = None)
+  (~run2_nanoAOD_ANY).toModify(proc.updatedJetsWithUserData.userFloats, vtxMass = None)
+  (~run2_nanoAOD_ANY).toModify(proc.updatedJetsWithUserData.userFloats, vtx3dL = None)
+  (~run2_nanoAOD_ANY).toModify(proc.updatedJetsWithUserData.userFloats, vtx3deL = None)
+  (~run2_nanoAOD_ANY).toModify(proc.updatedJetsWithUserData.userFloats, ptD = None)
+  (~run2_nanoAOD_ANY).toModify(proc.updatedJetsWithUserData.userFloats, genPtwNu = None)
```
Comment (kpedro88) [inline PhysicsTools/NanoAOD/python/custom_jme_cff.py:926]: you can make multiple modifications in a single `toModify()` statement (and therefore you should, in order to reduce duplication). example:
```python
(~run2_nanoAOD_ANY).toModify(proc.updatedJetsWithUserData.userFloats,
    leadTrackPt = None,
    leptonPtRel = None,
)
```

File: PhysicsTools/NanoAOD/python/custom_jme_cff.py [hunk 2]
```diff
@@ -1077,13 +1268,46 @@ def PrepJMECustomNanoAOD(process,runOnMC):
...
+      proc.nanoTableTaskCommon.add(proc.jetTask)
+      proc.nanoTableTaskCommon.add(proc.jetTablesTask)
+      proc.nanoTableTaskCommon.add(proc.jetForMETTask)
+    if addAK4PuppiJetTasks:
+      proc.nanoTableTaskCommon.add(proc.jetPuppiTask)
+      proc.nanoTableTaskCommon.add(proc.jetPuppiTablesTask)
+      proc.nanoTableTaskCommon.add(proc.jetPuppiForMETTask)
+    return proc
+
+  jmeNano_addAK4JetTaks_switch = cms.PSet(
```
Comment (kpedro88) [inline PhysicsTools/NanoAOD/python/custom_jme_cff.py:1298]: fix typo Taks -> Task

File: PhysicsTools/NanoAOD/python/custom_jme_cff.py (hunk 1 above)
Comment (nurfikri89) [reply PhysicsTools/NanoAOD/python/custom_jme_cff.py:926]: Thanks for the suggestion. Will clean up.

File: PhysicsTools/NanoAOD/python/custom_jme_cff.py (hunk 2 above)
Comment (nurfikri89) [reply PhysicsTools/NanoAOD/python/custom_jme_cff.py:1298]: To be corrected.

**Comment by nurfikri89** [comment]
> * 10224.15,11024.15,25202.15 which one is Run3 ?

I don't think any of them is Run-3.

> * What is the difference between the
>   FatJet and  FatJetForJEC
>   https://twiki.cern.ch/twiki/pub/CMS/JMECustomNanoAOD/JMENano_version4_forPR_39000_content.html#FatJetForJEC)
>   https://twiki.cern.ch/twiki/pub/CMS/JMECustomNanoAOD/JMENano_version4_forPR_39000_content.html#FatJet

`FatJet` is taken directly from slimmedJetsAK8 collection in MiniAOD with a minimum pt cut of 170 GeV while `FatJetForJEC` is a reclustered AK8 Puppi jet collection which can go as low as jet pt of 8 GeV with a basic set of jet variables saved for JEC studies.

**Comment by mariadalfonso** [comment]
> > * 10224.15,11024.15,25202.15 which one is Run3 ?
> 
> I don't think any of them is Run-3.

please test one of the Run3 locally

> > * What is the difference between the
> >   FatJet and  FatJetForJEC
> >   https://twiki.cern.ch/twiki/pub/CMS/JMECustomNanoAOD/JMENano_version4_forPR_39000_content.html#FatJetForJEC)
> >   https://twiki.cern.ch/twiki/pub/CMS/JMECustomNanoAOD/JMENano_version4_forPR_39000_content.html#FatJet
> 
> `FatJet` is taken directly from slimmedJetsAK8 collection in MiniAOD with a minimum pt cut of 170 GeV while `FatJetForJEC` is a reclustered AK8 Puppi jet collection which can go as low as jet pt of 8 GeV with a basic set of jet variables saved for JEC studies.

For large cone jets, do we really really need to go down to 8 GeV from 170 ?

**Comment by miquork** [comment]
It should be low enough if we go to 15 GeV for AK8 gen (FatJetForJEC). Reco jets matched to gen-jets of pT>15 GeV should still be kept down to zero. The rationale is staying sufficiently far from the underlying event, which is about 4 GeV per unit area for QCD and ttbar at gen level, with fluctuations of the order of sqrt(E). For AK4 this is pi*0.5^2 * 4 = 2 GeV of offset and sqrt(2)=1.4 GeV of noise, so 8 GeV is a safe 4 sigma above UE noise. For AK8 the corresponding numbers are x4 and x2 higher, so 8 GeV of UE with 3 GeV of noise. E.g. 15 GeV would be 2.5 sigma, 20 GeV 4 sigma from UE noise. I would still like to push this as low as possible to see where and how things break. E.g. SMP-HAD inclusive jets could measure AK8 this low.

**Comment by mariadalfonso** [comment]
please test

**Comment by mariadalfonso** [comment]
> It should be low enough if we go to 15 GeV for AK8 gen (FatJetForJEC). Reco jets matched to gen-jets of pT>15 GeV should still be kept down to zero. The rationale is staying sufficiently far from the underlying event, which is about 4 GeV per unit area for QCD and ttbar at gen level, with fluctuations of the order of sqrt(E). For AK4 this is pi*0.5^2 * 4 = 2 GeV of offset and sqrt(2)=1.4 GeV of noise, so 8 GeV is a safe 4 sigma above UE noise. For AK8 the corresponding numbers are x4 and x2 higher, so 8 GeV of UE with 3 GeV of noise. E.g. 15 GeV would be 2.5 sigma, 20 GeV 4 sigma from UE noise. I would still like to push this as low as possible to see where and how things break. E.g. SMP-HAD inclusive jets could measure AK8 this low.

ok, let's go to 15 for this.

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by nurfikri89** [comment]
Hi @srimanob, is there a Run-3 workflow which I can test with?

**Comment by srimanob** [comment]
test parameters:
-workflows = 10224.15,11024.15,25202.15

**Comment by srimanob** [comment]
Hi @nurfikri89 
Not yet for Run-3. You will need to modify
https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py#L1505-L1518
to accept Run-3, i.e. 
`return (fragment=="TTbar_13" or fragment=="TTbar_14TeV") and ('2017' in key or '2018' in key or '2021' in key) and ('FS' not in key)`

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by srimanob** [comment]
Hi @nurfikri89 @cms-sw/xpog-l2 
Maybe you should check if workflows with `.15` are OK. We can add Run-3 with the next commit, or in separate PR. I will retrigger the test, just to see if we can escape from INPUT problem which should not relate to this PR.

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by nurfikri89** [comment]
> Maybe you should check if workflows with `.15` are OK. We can add Run-3 with the next commit, or in separate PR. I will retrigger the test, just to see if we can escape from INPUT problem which should not relate to this PR.

I've re-run the `.15` workflows: 10224.15, 11024.15, 25202.15 and they are all okay.


> `return (fragment=="TTbar_13" or fragment=="TTbar_14TeV") and ('2017' in key or '2018' in key or '2021' in key) and ('FS' not in key)`

I've committed this change to my local branch. Should I push it to my remote branch now?

**Comment by srimanob** [comment]
Hi @nurfikri89 

Then you should test with Run-3 upgrade workflow `11634.15` runs OK or not,
`runTheMatrix.py --what upgrade -l 11634.15 --wm init`

If everything is OK, then please add also wf 11634.15 to 
`Configuration/PyReleaseValidation/python/relval_2017.py`. So we can trigger the test easily.

**Comment by srimanob** [comment]
test parameters:
- workflows = 10224.15,11024.15,25202.15,11634.15

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by srimanob** [comment]
@mariadalfonso 
Do you want to review and sign this PR? We can assign xpog here.

**Comment by srimanob** [comment]
@nurfikri89 
Just my curiosity, will JME nano be useful for FastSim somehow, i.e. study JEC in FastSim?

**Comment by nurfikri89** [comment]
> @nurfikri89 Just my curiosity, will JME nano be useful for FastSim somehow, i.e. study JEC in FastSim?

@srimanob  No plans so far. None that I know of.

**Comment by srimanob** [comment]
+Upgrade

The updated Run-2 and Run-3 JMENano workflows run fine. The PR does not include JMENano for FastSim so far. Not sure if it will be useful, @sbein. We can enable it later if needed.

**Comment by jpata** [comment]
+reconstruction
- POG code, no effect in standard reco

**Comment by kskovpen** [comment]
+pdmv

**Comment by nurfikri89** [comment]
My apologies for the late commits. There was a request to increase the precision of `rawFactor` jet variable. Additionally, made some changes in jetCollectionTools.py to ensure Puppi jets are reclustered using the puppi weights from the input MiniAOD.

**Comment by mariadalfonso** [comment]
please test workflow 11634.15

**Comment by mariadalfonso** [comment]
@nurfikri89 

in your previous attempt there was also this change
https://github.com/cms-sw/cmssw/pull/35614/files#diff-0c020df349ee3eaa784738e47d237d31c18691fe7925d2adb481e32962804f33

Is it still needed ? or was already fixed in another general PR ?

**Comment by nurfikri89** [comment]
> @nurfikri89
> 
> in your previous attempt there was also this change https://github.com/cms-sw/cmssw/pull/35614/files#diff-0c020df349ee3eaa784738e47d237d31c18691fe7925d2adb481e32962804f33
> 
> Is it still needed ? or was already fixed in another general PR ?

@mariadalfonso  It is fixed in this PR. We don't need `packedPrimaryVertexAssociationJME` anymore to run puppi.

**Comment by mariadalfonso** [comment]
> please test workflow 11634.15

Was the WF 11634.15 supposed to run by cmsbot ? I didn't found in the folder 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-12-2300+b27839/52113/

**Comment by srimanob** [comment]
Hi @mariadalfonso 

It runs, but not for comparison.

Here it is:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b27839/26794/runTheMatrix-results/11634.15_TTbar_14TeV+2021_JMENano+TTbar_14TeV_TuneCP5_GenSim+Digi+RecoNano+HARVESTNano+ALCA/

**Comment by mariadalfonso** [comment]
@nurfikri89

seems we have twice this branch
https://twiki.cern.ch/twiki/pub/CMS/JMECustomNanoAOD/JMENano_version4_forPR_39000_content.html#HLTriggerFinalPath

**Comment by nurfikri89** [comment]
> seems we have twice this branch https://twiki.cern.ch/twiki/pub/CMS/JMECustomNanoAOD/JMENano_version4_forPR_39000_content.html#HLTriggerFinalPath

@mariadalfonso  I could only find one branch with the name `HLTriggerFinalPath`. The other one is `HLTriggerFirstPath`.

**Comment by mariadalfonso** [comment]
+xpog

seems fine from technical point of view.
content wise not possible to verify the content w./o the DQM file
no confirmation that the 12B from Run2 UL were useful (AFIAK have never used to derive JEC)

to be done ASAP: 2 and 5 from here 
https://github.com/cms-sw/cmssw/pull/39000#issuecomment-1208526750
until then DO NOT set to central production

**Comment by jpata** [comment]
+reconstruction
- resign since https://github.com/cms-sw/cmssw/pull/39000#issuecomment-1213066883

**Comment by srimanob** [comment]
+Upgrade

From the upgrade side, this PR update the workflows for JMENano. Workflow runs fine. For technical topics, i.e. contents or speed up, please follow up with @cms-sw/xpog-l2.

**Comment by mariadalfonso** [comment]
@nurfikri89 please backport to 12_4

**Comment by kskovpen** [comment]
+pdmv

**Comment by perrotta** [comment]
+1
- As far as I understand from https://github.com/cms-sw/cmssw/pull/39000#issuecomment-1214946228, this has to be further verified before backporting it in the nanoAOD production release 12_4_X: next week 12_5_0_pre5 relvals could be a good opportunity to do so

**Comment by srimanob** [comment]
@mariadalfonso @cms-sw/pdmv-l2 @nurfikri89 
Just to confirm on the content validation, should the workflows with .15 be submitted with the next pre-release to get DQM for validation? Or this will be done from x-pog or JME side privately. Thanks.

**Comment by mariadalfonso** [comment]
> @mariadalfonso @cms-sw/pdmv-l2 @nurfikri89
> Just to confirm on the content validation, should the workflows with .15 be submitted with the next pre-release to get DQM for validation? Or this will be done from x-pog or JME side privately. Thanks.

The DQM was just never in the release. 
We need something like this
https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/python/nanoDQM_cfi.py

Once this is added we can aim to test it: 
* JMEnano vs JMEnano
* JMEnano vs central nano (w.o this pretty much I'm not sure what we use for calibration is what is going to be used for analysis given the large amount of cooking in the JMEnano).


The 12B of UL dataset were submitted by PPD for FWK validation and BLINDLY.  Also all the presentation shown at PPD never included a single plot and nd PRs desciption never contain distributions.
Still waiting to close the loop on this unfortunately.

**Pull request merged**